src/controller/worker.rs

```diff
@@ -0,0 +1,161 @@
+        .unwrap_or(5);
+}
+
+pub struct Worker<A: AllocationsChecker> {
+    state_manager: Arc<StateManager>,
+    allocations_checker: Arc<A>,
+    queries_tx: mpsc::Sender<QueryTask>,
+    queries_rx: UseOnce<mpsc::Receiver<QueryTask>>,
+}
+
+pub struct QueryTask {
+    pub dataset: Dataset,
+    pub query_str: String,
+    pub client_id: Option<PeerId>,
+    pub response_sender: oneshot::Sender<Result<QueryResult, QueryError>>,
+}
+
+impl<A: AllocationsChecker> Worker<A> {
+    pub fn new(state_manager: Arc<StateManager>, allocations_checker: Arc<A>) -> Self {
+        let (queries_tx, queries_rx) = mpsc::channel(*QUEUED_QUERIES);
+        Self {
+            state_manager,
+            allocations_checker,
+            queries_tx,
+            queries_rx: UseOnce::new(queries_rx),
+        }
+    }
+
+    pub fn set_desired_chunks(&self, chunks: ChunkSet) {
+        self.state_manager.set_desired_chunks(chunks);
+    }
+
+    pub fn set_datasets_index(&self, datasets_index: DatasetsIndex) {
+        self.state_manager.set_datasets_index(datasets_index);
+    }
+
+    pub fn stop_downloads(&self) {
+        self.state_manager.stop_downloads();
+    }
+
+    pub fn status(&self) -> manager::Status {
+        self.state_manager.current_status()
+    }
+
+    pub fn schedule_query(
+        &self,
+        query_str: String,
+        dataset: Dataset,
+        client_id: Option<PeerId>,
+    ) -> Option<impl Future<Output = Result<QueryResult, QueryError>> + '_> {
+        let (resp_tx, resp_rx) = oneshot::channel();
+        match self.queries_tx.try_send(QueryTask {
+            dataset,
+            query_str,
+            client_id,
+            response_sender: resp_tx,
+        }) {
+            Err(mpsc::error::TrySendError::Full(_)) => {
+                return None;
+            }
+            Err(mpsc::error::TrySendError::Closed(_)) => {
+                panic!("Query subscriber dropped");
+            }
+            Ok(_) => {
+                metrics::PENDING_QUERIES.inc();
+            }
+        };
+        Some(async move {
+            resp_rx
+                .await
+                .expect("Query processor didn't produce a result")
+        })
+    }
+
+    pub async fn run(&self, cancellation_token: CancellationToken) {
+        let queries_rx = self.queries_rx.take().unwrap();
+        ReceiverStream::new(queries_rx)
+            .take_until(cancellation_token.cancelled_owned())
+            .for_each_concurrent(*PARALLEL_QUERIES, |query_task| async move {
+                metrics::PENDING_QUERIES.dec();
+                tracing::debug!(
+                    "Running query from {}",
+                    query_task
+                        .client_id
+                        .map(|id| id.to_string())
+                        .unwrap_or("{unknown}".to_string())
+                );
+                let result = match self
+                    .allocations_checker
+                    .try_spend(query_task.client_id)
```

Comment:
Could this possibly be done in the `handle_query` function? This would decouple `Worker` from something that's essentially internal to P2P transport. We could remove the type parameter then.

Reply:
Yeah, I also wanted to do it (as mentioned in the PR description), but it wasn't that easy. I can make it in a separate commit

Reply:
OK, just leave a TODO comment maybe.